astroid/interpreter/objectmodel.py

```diff
@@ -119,17 +121,41 @@ def lookup(self, name):
         raise AttributeInferenceError(target=self._instance, attribute=name)
 
     @property
-    def attr___new__(self):
-        """Calling cls.__new__(cls) on an object returns an instance of that object.
+    def attr___new__(self) -> bases.BoundMethod:
+        """Calling cls.__new__(type) on an object returns an instance of 'type'."""
+        node: nodes.FunctionDef = builder.extract_node(
+            """def __new__(self, cls): return cls()"""
+        )
+        # We set the parent as being the ClassDef of 'object' as that
+        # triggers correct inference as a call to __new__ in bases.py
+        node.parent: nodes.ClassDef = AstroidManager().builtins_module["object"]
+
+        # TODO: Use isinstance instead of try ... except after _instance has typing
+        try:
+            bound = self._instance._proxied
+        except AttributeError:
+            bound = self._instance
+        return bases.BoundMethod(proxy=node, bound=bound)
+
+    @property
+    def attr___init__(self) -> bases.BoundMethod:
+        """Calling cls.__init__() normally returns None."""
+        # The *args and **kwargs are necessary not too trigger warnings about missing
+        # or extra parameters for '__init__' methods we don't infer correctly.
+        # This BoundMethod is the fallback value for those.
+        node: nodes.FunctionDef = builder.extract_node(
+            """def __init__(self, *args, **kwargs): return None"""
+        )
+        # We set the parent as being the ClassDef of 'object' as that
+        # is where this method originally comes from
+        node.parent: nodes.ClassDef = AstroidManager().builtins_module["object"]
 
-        Instance is either an instance or a class definition of the instance to be
-        created.
-        """
         # TODO: Use isinstance instead of try ... except after _instance has typing
         try:
-            return self._instance._proxied.instantiate_class()
+            bound = self._instance._proxied
         except AttributeError:
-            return self._instance.instantiate_class()
+            bound = self._instance
+        return bases.BoundMethod(proxy=node, bound=bound)
```

Comment:
This looks the same as for `attr___new__`: how would you feel about factoring it out?

Reply:
I'll add a little helper function.